lib/straight/gateway.rb

```diff
@@ -90,32 +89,33 @@ def new_order(args)
         if args[:amount].nil? || !args[:amount].kind_of?(Numeric) || args[:amount] <= 0
           raise OrderAmountInvalid, "amount cannot be nil and should be more than 0"
         end
-        # Setting default values
-        args[:currency]         ||= default_currency
-        args[:btc_denomination] ||= :satoshi
 
-        amount = args[:amount_from_exchange_rate] = amount_from_exchange_rate(
+        # Build order with default values
+        order = Kernel.const_get(order_class).new
+        order.gateway = self
+        order.keychain_id = args[:keychain_id]
+        order.block_height_created_at = fetch_latest_block_height rescue nil
+
+         args[:currency] ||= default_currency
+
+        args[:amount_from_exchange_rate] = amount_from_exchange_rate(
           args[:amount],
-          currency:         args[:currency],
-          btc_denomination: args[:btc_denomination]
+          currency: args[:currency],
+          btc_denomination: args[:btc_denomination] || :satoshi,
         )
```

Comment:
If you have started to refactor:
1) Changing parameters (hash) inside functions is bad practice because hash is mutable and all your changes will be applied for it outside function.
2) Why did you write `btc_denomination: args[:btc_denomination] || :satoshi`, but `args[:currency] ||= default_currency` has been left? As for me, code before your commits was better than now, but not perfect. You should rework your code to make it looks more elegant or you could leave it as it is.